src/backends/r1cs/mod.rs

```diff
@@ -0,0 +1,557 @@
+    fn finalize_circuit(
+        &mut self,
+        public_output: Option<crate::var::Var<Self::Field>>,
+        returned_cells: Option<Vec<CellVar>>,
+        _private_input_indices: Vec<usize>,
+        _main_span: crate::constants::Span,
+    ) -> crate::error::Result<()> {
+        // store the return value in the public input that was created for that
+        if let Some(public_output) = public_output {
+            let cvars = &public_output.cvars;
+
+            for (pub_var, ret_var) in cvars.clone().iter().zip(returned_cells.unwrap()) {
+                // replace the computation of the public output vars with the actual variables being returned here
+                let var_idx = pub_var.idx().unwrap();
+                let prev = self
+                    .witness_vars
+                    .insert(var_idx, Value::PublicOutput(Some(ret_var)));
+                assert!(prev.is_some());
+            }
+        }
+
+        // for sanity check, we make sure that every cellvar created has ended up in a r1cs constraint
+        let mut written_vars = HashSet::new();
+        for constraint in &self.constraints {
+            for lc in constraint.as_array() {
+                if let Some(terms) = &lc.terms {
+                    for var in terms.keys() {
+                        written_vars.insert(var.index);
+                    }
+                }
+            }
+        }
+
+        // check if every cell vars end up being a cell var in the circuit or public output
+        for index in 0..self.next_variable {
+            if !written_vars.contains(&index) {
+                // check if outputs contains the cell var that has the same index
+                if !self.public_outputs.iter().any(|&x| x.index == index) {
+                    panic!("there's a bug in the circuit_writer, some cellvar does not end up being a cellvar in the circuit!");
+                }
+            }
+        }
+
+        self.finalized = true;
+
+        Ok(())
+    }
+
+    /// Generate the witnesses 
+    /// This process should check if the constraints are satisfied.
+    fn generate_witness(
+        &self,
+        witness_env: &mut crate::witness::WitnessEnv<Self::Field>,
+    ) -> crate::error::Result<Self::GeneratedWitness> {
+        if !self.finalized {
+            panic!("the circuit is not finalized yet!");
+        }
+
+        let mut witness = HashMap::<usize, Fr>::new();
+
+        for constraint in &self.constraints {
+            for lc in &constraint.as_array() {
+                if let Some(terms) = &lc.terms {
+                    for var in terms.keys() {
+                        if witness.contains_key(&var.index) {
+                            continue;
+                        }
+                        let val = self.compute_var(witness_env, *var)?;
+                        witness.insert(var.index, val);
+                    }
+                }
+            }
```

Comment:
I'm not sure I follow why we do this here. Why can't we just do something like

```rust
let witness = witness.iter().map(|v| self.compute_val(witness_env, v)).collect_vec();
```

although:

* you'd have to skip the public output and compute it at the end
* you'd have to have a `compute_val` function that implements this logic:

<img width="385" alt="Screenshot 2024-05-01 at 11 41 30 AM" src="https://github.com/zksecurity/noname/assets/1316043/5678475d-1bc8-41de-b308-3f97eae1b578">

Reply:
the whole idea is that:

* the witness vector is a single vector
* the index in that vector can be the same as the index of a CellVar
* when we compile the circuit, we know exactly when we create new variables by pushing in that witness vector
* so we can have a "symbolic" witness vector that tells us how to compute each cell of that vector at runtime

Reply:
I am wondering if we should have the `witness_vars` as `Vec<(CellVar, Value)>`, instead of just `Vec<Value>`.

The `compute_var` require the argument in CellVar type, while the other places require `Value`.

With both stored in the vector, we can do the following without the need to retrieve the cell vars from the linear combinations in constraints.
```
let witness = self.witness_vars.iter().map(|v| self.compute_val(witness_env, v.0)).collect_vec();
```

Reply:
but here we are only interested in the index in CellVar here, which is going to be 0, 1, 2, 3, etc.

Reply:
and we need to recreated the CellVar with a span in order to pass into the `compute_var`, right?

Reply:
I was thinking that compute_var is overkill here, we can export the match statement inside of it and have it live in another fn

Reply:
Ah, so we will need to refactor the `compute_var` to call `compute_val` which encapsulate the match statements for `symbolic vars` (the Value type)

Reply:
https://github.com/zksecurity/noname/issues/60

Reply:
lmk if this is correct way: 530d2b47a2b59b3e26dca2ed1d6a165b33b02a09

Reply:
LGTM!